server.go

```diff
@@ -462,11 +462,85 @@ func (p *sshFxpOpenPacket) respond(svr *Server) responsePacket {
 		osFlags |= os.O_EXCL
 	}
 
-	f, err := os.OpenFile(svr.toLocalPath(p.Path), osFlags, 0o644)
+	name := svr.toLocalPath(p.Path)
+	mode := os.FileMode(0o644)
+
+	var applyAttrs []func(*os.File) error
+
+	// Both `sshFileXferAttrPermissions` and `sshFileXferAttrACmodTime` are set
+	// by e.g. `sftp`. Just in case, we handle all other cases as well.
+	// Only apply for newly created files.
+	var err error
+	useAttrs := true
+	if _, err := os.Stat(name); err == nil {
+		useAttrs = false
+	}
+	if b, ok := p.Attrs.([]byte); useAttrs && ok {
+		if (p.Flags & sshFileXferAttrSize) != 0 {
+			var size uint64
+			if size, b, err = unmarshalUint64Safe(b); err == nil {
+				applyAttrs = append(applyAttrs, func(f *os.File) error {
+					return f.Truncate(int64(size))
+				})
+			}
+		}
+		if err != nil {
+			return statusFromError(p.ID, err)
+		}
+		if (p.Flags & sshFileXferAttrUIDGID) != 0 {
+			var uid uint32
+			var gid uint32
+			if uid, b, err = unmarshalUint32Safe(b); err != nil {
+			} else if gid, b, err = unmarshalUint32Safe(b); err != nil {
+			} else {
+				applyAttrs = append(applyAttrs, func(f *os.File) error {
+					return f.Chown(int(uid), int(gid))
+				})
+			}
+		}
+		if err != nil {
+			return statusFromError(p.ID, err)
+		}
+		if (p.Flags & sshFileXferAttrPermissions) != 0 {
+			var attrMode uint32
+			if attrMode, b, err = unmarshalUint32Safe(b); err == nil {
+				// Optionally, we could apply umask here.
+				mode = os.FileMode(attrMode)
```

Comment:
This conversion is not actually valid. These two types disagree on quite a number of bits.

Reply:
Isn't the same conversion being done in the newer implementations as well? https://github.com/pkg/sftp/blob/5bdc2b0e679d0d8336dc63f8b790108667ea4280/internal/encoding/ssh/filexfer/attrs.go#L186

Reply:
We’ll want to wrap all of these through either [toFileMode](https://github.com/pkg/sftp/blob/master/stat_posix.go#L53) or [fromFileMode](https://github.com/pkg/sftp/blob/master/stat_posix.go#L87)

Reply:
While looking at the OpenSSH implementation, I saw that OpenSSH also has no handling and uses the mode as-is.

Reply:
OK, so, this is kind of weird to explain. Yes, OpenSSH uses the mode directly, because the SFTP standard is that the field conforms to the “POSIX filemode bit assignment”.

However, _Go_ doesn’t use the POSIX filemode bit assignment. It uses a subtly different one, which is why we have the `toFileMode` and `fromFileMode` functions in the first place.

Reply:
Oh, TIL. That's good context, thanks for sharing!